File: max-product-subarray/maximumProductSubarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        maxHerePre = nums[0]
        minHerePre = nums[0]
        maxSoFar = nums[0]
        
        for i in range(1,len(nums)):
            maxHere = max(max(maxHerePre * nums[i], minHerePre * nums[i]), nums[i])
            minHere = min(min(maxHerePre * nums[i], minHerePre * nums[i]), nums[i])
            maxSoFar = max(maxHere, maxSoFar)
            maxHerePre = maxHere
            minHerePre = minHere
            
        return maxSoFar

File: max-product-subarray/maximumProductSubarray.go
func max(a, b int) int{
    if a < b {
        return b
    }
    return a
}

func min(a, b int) int{
    if a < b {
        return a
    }
    return b
}

func maxProduct(nums []int) int {
    var maxHerePre int = nums[0]
    var minHerePre int = nums[0]
    var maxSoFar int = nums[0]
    var maxHere, minHere int
    for i := 1; i < len(nums); i++ {
        maxHere = max(max(maxHerePre * nums[i], minHerePre * nums[i]), nums[i]);
        minHere = min(min(maxHerePre * nums[i], minHerePre * nums[i]), nums[i]);
        maxSoFar = max(maxSoFar, maxHere)
        maxHerePre = maxHere
        minHerePre = minHere
    }
    return maxSoFar
}
